### DISCUSSION TITLE
Why `/zip` is read-only?

### DISCUSSION THREAD

**Discussion by franalbani** [jart/cosmopolitan#1308] [Q&A]
For some reason, I was confident that it was possible for a cosmo binary to write inside `/zip` (and even persist the changes in the file itself).

But after failing to do it with cosmo's python:

```console
$ ./python -c "open('/zip/memories.txt', 'w').write('memory')"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
OSError: [Errno 30] Read-only file system: '/zip/memories.txt'
```

and then verifying with a simple C example, I started `git grep`ing the repo searching for something to prove myself I was wrong.

[This is](https://github.com/jart/cosmopolitan/blob/85c58be94279f9a6b8d6e57ca13d3b4912a82042/libc/runtime/zipos-open.c#L208) the only thing I found:

> // zipos being a read-only filesystem, ...

I would like to understand why it is so and if there is another way to achieve the same result.

Thanks!

**Comment by metaist** [comment]
[`cosmofy`](https://github.com/metaist/cosmofy) is a tool I wrote that has some [lower-level tools](https://github.com/metaist/cosmofy?#cosmofy-fs) that let you `ls/cat/add/rm` on a cosmo python file.

I have run this tool on itself, so, yes, it is possible for a cosmo file to write to itself. You just have to remember to use `zipfile.ZipFile` instead of `open()` because you're technically modifying a zip file, not just writing bytes to disk.

**Reply by franalbani** [reply]
Nice!

I ignored that.

I leave this for future readers:

```console
$ wget https://cosmo.zip/pub/cosmos/v/4.0.2/bin/python
$ chmod +x python
$ ./python -c 'from zipfile import ZipFile; ZipFile("python", "a").open("example.txt", "w").write(b"something")'
$ ./python -c 'print(open("/zip/example.txt").read())'
something
```

**Reply by couleurm** [reply]
what would the C equivalent look like?
